<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Detection of Differentially Interacting Chromatin Regions From Multiple Hi-C Datasets • HiCcompareWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Detection of Differentially Interacting Chromatin Regions From Multiple Hi-C Datasets">
<meta property="og:description" content="This workshop will introduce methods for the comparative 
    (aka differential) analysis of the three-dimensional (3D) structure 
    of the genome using data generated by high-throughput chromatin 
    conformation capture (Hi-C) technologies. Hi-C data allows for insights 
    into the genome-wide 3D genomic interactions which play an important role 
    in regulating gene expression and other genomic processes. Just as 
    differential expression analyses using RNA-seq data have become 
    a routine part of genomic experiments, we expect the differential analysis 
    of genomic interactions to become a common task. This workflow will help 
    a novice learn to perform differential analysis of two or more Hi-C datasets 
    and interpret the results of the differential genomic interactions.  ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">HiCcompareWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/hic_tutorial.html">Detection of Differentially Interacting Chromatin Regions From Multiple Hi-C Datasets</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mdozmorov/HiCcompareWorkshop/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<p><img src="https://github.com/mdozmorov/HiCcompareWorkshop/workflows/.github/workflows/basic_checks.yaml/badge.svg" alt=".github/workflows/basic_checks.yaml"></p>
<div id="detection-of-differentially-interacting-chromatin-regions-from-multiple-hi-c-datasets" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#detection-of-differentially-interacting-chromatin-regions-from-multiple-hi-c-datasets" class="anchor"></a>Detection of Differentially Interacting Chromatin Regions From Multiple Hi-C Datasets</h1></div>
<p><a href="https://github.com/mdozmorov/HiCcompareWorkshop">HiCcompareWorkshop</a> introduces methods for the comparative (aka differential) analysis of the three-dimensional (3D) structure of the genome using data generated by high-throughput chromatin conformation capture (Hi-C) technologies. Hi-C data allows for insights into the genome-wide 3D genomic interactions which play an important role in regulating gene expression and other genomic processes. Just as differential expression analyses using RNA-seq data have become a routine part of genomic experiments, we expect the differential analysis of genomic interactions to become a common task. This workflow will help a novice learn to perform differential analysis of two or more Hi-C datasets and interpret the differential genomic interactions’ results.</p>
<p>This workshop is based on <a href="https://currentprotocols.onlinelibrary.wiley.com/doi/abs/10.1002/cpbi.76">Stansfield et al., “R Tutorial: Detection of Differentially Interacting Chromatin Regions From Multiple Hi-C Datasets.”</a>.</p>
<ul>
<li><a href="https://mdozmorov.github.io/Talk_HiCcompare/index.html">Slides with a brief introduction for the workshop</a></li>
<li><a href="https://github.com/mdozmorov/HiCcompareWorkshop">HiCcompareWorkshop GitHub repo</a></li>
<li><a href="https://hub.docker.com/repository/docker/mdozmorov/hiccompareworkshop">HiCcompareWorkshop Docker image</a></li>
<li><a href="https://mdozmorov.github.io/HiCcompareWorkshop/">HiCcompareWorkshop pkgdown website</a></li>
</ul>
<p>This workshop will be presented at the <a href="https://bioc2020.bioconductor.org/">Bioconductor Virtual Conference 2020</a>, July 30, 2020, 10:00-10:55am</p>
</div>
<div id="using-docker" class="section level1">
<h1 class="hasAnchor">
<a href="#using-docker" class="anchor"></a>Using Docker</h1>
<p>The easiest way to get started with the workshop is to run it from a Docker container.</p>
<ul>
<li>Run <code>docker run -e PASSWORD=yourpassword -p 8787:8787 -d --rm mdozmorov/hiccompareworkshop</code>. Use <code>-v $(pwd):/home/rstudio</code> argument to map your local directory to the container.</li>
<li>Log in to RStudio at <a href="http://localhost:8787">http://localhost:8787</a> using username <code>rstudio</code> and password <code>yourpassword</code>. Note that on Windows you need to provide your localhost IP address like <code>http://191.163.92.108:8787/</code> - find it using <code>docker-machine ip default</code> in Docker’s terminal.</li>
<li>Run <code><a href="https://rdrr.io/r/utils/browseVignettes.html">browseVignettes(package = "HiCcompareWorkshop")</a></code>. Click on one of the links, “HTML”, “source”, “R code”.
<ul>
<li>In case of <code>The requested page was not found</code> error, add <code>help/</code> to the URL right after the hostname, e.g., <a href="http://localhost:8787/help/library/HiCcompareWorkshop/doc/hic_tutorial.html">http://localhost:8787/help/library/HiCcompareWorkshop/doc/hic_tutorial.html</a>. This is a <a href="https://github.com/rocker-org/rocker-versioned/issues/178">known bug</a>.</li>
</ul>
</li>
</ul>
</div>
<div id="local-installation" class="section level1">
<h1 class="hasAnchor">
<a href="#local-installation" class="anchor"></a>Local Installation</h1>
<pre><code>if(!require(devtools)) install.packages("devtools")
devtools::install_github(repo = "mdozmorov/HiCcompareWorkshop", build_vignettes = TRUE)</code></pre>
<p>If installation fails due to missing packages, install them as follows:</p>
<pre><code>if(!require(BiocManager)) install.packages("BiocManager")
BiocManager::install(c('edgeR', 'HiCcompare', 'multiHiCcompare', 'clusterProfiler', 'ROntoTools'))</code></pre>
</div>
<div id="workshop-description" class="section level1">
<h1 class="hasAnchor">
<a href="#workshop-description" class="anchor"></a>Workshop Description</h1>
<div id="instructor-name-and-contact-information" class="section level2">
<h2 class="hasAnchor">
<a href="#instructor-name-and-contact-information" class="anchor"></a>Instructor name and contact information</h2>
<ul>
<li>Mikhail Dozmorov (<a href="mailto:mikhail.dozmorov@vcuhealth.org" class="email">mikhail.dozmorov@vcuhealth.org</a>)</li>
</ul>
</div>
<div id="pre-requisites" class="section level2">
<h2 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h2>
<ul>
<li>Basic knowledge of R syntax and command-line tools</li>
<li>Familiarity with Hi-C chromatin conformation capture technology</li>
<li>Understanding of Hi-C data properties (distance-dependent decay of interaction frequencies, biases)</li>
<li>Familiarity with specialized (<code>.hic</code> <a href="http://aidenlab.org/data.html">http://aidenlab.org/data.html</a>, <code>.cool</code> <a href="ftp://cooler.csail.mit.edu/coolers">ftp://cooler.csail.mit.edu/coolers</a>) and text-based (sparse upper-triangular, full square matrix) Hi-C data formats is desirable</li>
</ul>
</div>
<div id="r--bioconductor-packages-used" class="section level2">
<h2 class="hasAnchor">
<a href="#r--bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h2>
<ul>
<li><a href="https://www.bioconductor.org/packages/HiCcompare">HiCcompare</a></li>
<li><a href="https://bioconductor.org/packages/multiHiCcompare">multiHiCcompare</a></li>
</ul>
</div>
<div id="time-outline" class="section level2">
<h2 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h2>
<p>The workshop duration is 55 min. Approximate timing of activities:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Overview</td>
<td>5m</td>
</tr>
<tr class="even">
<td>Data representation and manipulation</td>
<td>10m</td>
</tr>
<tr class="odd">
<td>Differential analysis of Hi-C data</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Interpretation of Hi-C differences</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>Questions and answers session</td>
<td>15m</td>
</tr>
</tbody>
</table>
<div id="learning-goals" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-goals" class="anchor"></a>Learning Goals</h3>
<ul>
<li>Get familiar with Hi-C data import into R</li>
<li>Understand the visualization of biases between pairs of Hi-C datasets using Mean-Distance (MD) plot</li>
<li>Learn the Loess-based normalization strategy that minimizes between-dataset differences</li>
<li>Perform differential analysis of <span class="math inline">2 × 2</span> groups of Hi-C data, accounting for covariates, if needed</li>
<li>Visualize and understand interaction frequency differences</li>
<li>Get familiar with approaches for interpretation of chromatin interaction differences in the context of genes and gene expression differences</li>
</ul>
</div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/mdozmorov/HiCcompareWorkshop/">https://​github.com/​mdozmorov/​HiCcompareWorkshop/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/mdozmorov/HiCcompareWorkshop/issues/new/choose">https://​github.com/​mdozmorov/​HiCcompareWorkshop/​issues/​new/​choose</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Mikhail Dozmorov <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-0086-8358" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Mikhail Dozmorov.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
