---
output:
  rmarkdown::html_document:
    highlight: pygments
    toc: true
    toc_depth: 3
    fig_width: 5
# bibliography: "`r system.file(package='dummychapter1', 'vignettes', 'bibliography.bib')`"
vignette: >
  %\VignetteIndexEntry{HiCtutorial}
  %\VignetteEncoding[utf8]{inputenc}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

# Detection of Differentially Interacting Chromatin Regions From Multiple Hi-C Datasets

## Workshop information

### Instructor(s) name(s) and contact information

Mikhail Dozmorov

Department of Biostatistics  
Virginia Commonwealth University  
830 E. Main St  
Richmond, VA, 23298  
mikhail.dozmorov@vcuhealth.org  

### Workshop Description

This tutorial will introduce methods for the comparative (aka differential) analysis of the three-dimensional (3D) structure of the genome using data generated by high throughput chromatin conformation capture (Hi-C) technologies. Hi-C data allows for insights into the genome-wide genomic interactions which play an important role in the regulation of the genome. Just as differential expression analyses using RNA-seq data have become a routine part of genomic experiments, we expect the differential analysis of genomic interactions to become a common task for a bioinformatician. This workflow will help a novice learn to perform differential analysis of two or more Hi-C datasets and interpret the results of the differential genomic interactions.  

### Pre-requisites

* Basic knowledge of R syntax and command-line tools
* Familiarity with Hi-C chromatin conformation capture technology and FASTQ format
* Understanding of Hi-C data properties (distance-dependent decay of interaction frequencies, biases)
* Familiarity with specialized (`.hic` [http://aidenlab.org/data.html](http://aidenlab.org/data.html), `.cool` [ftp://cooler.csail.mit.edu/coolers](ftp://cooler.csail.mit.edu/coolers)) and text-based (sparse upper-triangular, full square matrix) Hi-C data formats

### Workshop Participation

Execution of example code and hands-on practice

### _R_ / _Bioconductor_ packages used

* [HiCcompare](https://www.bioconductor.org/packages/HiCcompare)
* [multiHiCcompare](https://bioconductor.org/packages/multiHiCcompare)


### Time outline

| Activity                                              | Time |
|-------------------------------------------------------|------|
| Overview                                              | 5m   |
| Data representation and manipulation                  | 20m  |
| Differential analysis of Hi-C data                    | 20m  |
| Interpretation of Hi-C differences                    | 15m  |


### Learning Goals

* Get familiar with Hi-C data import into R
* Understand the visualization of biases between pairs of Hi-C datasets using Mean-Distance (MD) plot
* Learn the Loess-based normalization strategy that minimizes between-dataset differences
* Perform differential analysis of 2X2 groups of Hi-C data
* Visualize and understand interaction frequency differences
* Learn how the significance of differential interactions changes with the increasing distance between interacting regions
* Get familiar with approaches to interpret chromatin interaction differences in context of genes

```{r setup, echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
# Set up the environment
opts_chunk$set(cache.path='cache/', fig.path='img/', cache=T, tidy=F, fig.keep='high', echo=T, dpi=100, warnings=F, message=F, comment=NA, warning=F, results='as.is', fig.width = 10, fig.height = 6) #out.width=700, 
set.seed(1)
options(stringsAsFactors = FALSE)
```

### Working with processed Hi-C data

The `HiCcompare` R package was designed for working with processed Hi-C data. It contains several functions which may be useful for an analysis. Hi-C data extracted from `.hic` files using `straw` can be simply read into R in the standard fashion for loading any text file containing data. `HiCcompare` can then be used to convert Hi-C data from sparse upper triangular matrix format into a full contact matrix using the `sparse2full` function. This process can be reversed using the `full2sparse` function. Data aligned by `HiC-Pro` can be converted into a more usable `BEDPE` format using the `hicpro2bedpe` function. The `hicpro2bedpe` function takes the `.matrix` and `.bed` files produced by `HiC-Pro` as input and produces a sparse upper triangular matrix containing start and end coordinates for each interacting region. 

The original `HiCCompare` R package can be used when only two Hi-C datasets are available to be compared [@Stansfield2018]. `HiCcompare` provides a method for the joint normalization and difference detection of two Hi-C datasets, but cannot be generalized to higher numbers of datasets. `multiHiCcompare` will need to be used if more than two Hi-C datasets are to be compared [@Stansfield:multihiccompare].

### Obtaining and preparing the data

We begin our tutorial illustrating the processing of Hi-C datasets to prepare them for comparative analysis. First, we will need to download an example set of Hi-C data. We will use data from Rao 2017 [@Rao:2017aa]. For simplicity, we will only use two replicates for each experimental condition. The experimental conditions are normal HCT-116 cells and HCT-116 cells treated with auxin for six hours. To download the `.hic` files from GEO run the following commands in the terminal. Note: downloading the data for this protocol will require about 30GB of hard drive space. This step is for illustration purposes only, we will work with the much smaller data already processed using these commands.

```{bash eval=FALSE}
wget https://github.com/theaidenlab/straw/tree/master/bin/Mac/straw
chmod +x straw

wget ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM2795nnn/GSM2795535/suppl/GSM2795535_Rao-2017-HIC001_30.hic
wget ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM2795nnn/GSM2795536/suppl/GSM2795536_Rao-2017-HIC002_30.hic
wget ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM2809nnn/GSM2809539/suppl/GSM2809539_Rao-2017-HIC008_30.hic
wget ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM2809nnn/GSM2809540/suppl/GSM2809540_Rao-2017-HIC009_30.hic

# Make directories for the contact map files
mkdir HIC001
mkdir HIC002
mkdir HIC008
mkdir HIC009

# Extract contact maps using straw by running the following commands in the terminal
# Or, put the commands into a script file, e.g., `straw.sh`, and run it
for i in {1..22}
    do
        ./straw NONE GSM2795535_Rao-2017-HIC001_30.hic $i $i BP 500000 > HIC001/HIC001.NONE.chr$i.500000.txt
    done
    ./straw NONE GSM2795535_Rao-2017-HIC001_30.hic X X BP 500000 > HIC001/HIC001.NONE.chrX.500000.txt
    
for i in {1..22}
    do
        ./straw NONE GSM2795536_Rao-2017-HIC002_30.hic $i $i BP 500000 > HIC002/HIC002.NONE.chr$i.500000.txt
    done
    ./straw NONE GSM2795536_Rao-2017-HIC002_30.hic X X BP 500000 > HIC002/HIC002.NONE.chrX.500000.txt
    
for i in {1..22}
    do
        ./straw NONE GSM2809539_Rao-2017-HIC008_30.hic $i $i BP 500000 > HIC008/HIC008.NONE.chr$i.500000.txt
    done
    ./straw NONE GSM2809539_Rao-2017-HIC008_30.hic X X BP 500000 > HIC008/HIC008.NONE.chrX.500000.txt
        
for i in {1..22}
    do
        ./straw NONE GSM2809540_Rao-2017-HIC009_30.hic $i $i BP 500000 > HIC009/HIC009.NONE.chr$i.500000.txt
    done
    ./straw NONE GSM2809540_Rao-2017-HIC009_30.hic X X BP 500000 > HIC009/HIC009.NONE.chrX.500000.txt
```

These steps will create four folders containing the sparse upper triangular matrices for chromosomes 1-22, and X for each sample. HIC001 and HIC002 are the two replicates for the normal HCT-116 cells and HIC008 and HIC009 are the two replicates for the auxin-treated HCT-116 cells. 

We now need to read the data into R. Open R and make sure the working directory is set to the directory where the Hi-C data is stored and execute the following commands:

```{r, message=FALSE}
# Install, if necessary, and load necessary libraries and set up R session
if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
library(readr) # install.packages("readr")
library(data.table) # install.packages("data.table")
library(dplyr) # install.packages("dplyr")
library(edgeR) # BiocManager::install("edgeR")
library(BiocParallel) # BiocManager::install("BiocParallel") 
library(HiCcompare) # BiocManager::install("HiCcompare"), or, for the latest version, devtools::install_github('dozmorovlab/HiCcompare', build_vignettes = TRUE, force = TRUE)
# install.packages("devtools")
library(multiHiCcompare) # BiocManager::install("multiHiCcompare", version = "devel") 
options(scipen = 10) # Output fixed numbers, not scientific notation

# Set up parameters for reading in data
chr <- paste0('chr', c(1:22, 'X')) # Chromosome names
chr <- paste0('chr', c(1)) # Chromosome names
dirIn <- "data/" # Input folder
dirIn <- "https://github.com/kmt555/ComparativeHiC-analysis-tutorial/raw/master/data/" # Remote data
samples <- paste0('HIC00', c(1, 2, 8, 9)) # Sample names
res <- 500000 # Data resolution

# Read data
sample_list <- list()
chr_list    <- list()
for( j in 1:length(samples)) {
  for (i in 1:length(chr)) {
    chr_list[[i]] <- read_tsv(paste0(dirIn, samples[j], "/", samples[j], 
                                     ".NONE.", chr[i], ".", res, ".txt.gz"), 
                              col_names = FALSE) %>% as.data.table()
    # Add column indicating the chromosome
    chr_list[[i]] <- cbind(i, chr_list[[i]]) 
    colnames(chr_list[[i]]) <- c('chr', 'region1', 'region2', 'IF')
  }
  sample_list[[j]] <- chr_list
  chr_list <- list()
}

# Collapse separate chromosome lists into one table per sample
sample_list <- lapply(sample_list, rbindlist)
```

We now have a list with each entry containing the sparse upper triangular matrix for one of the Hi-C datasets:

```{r}
sample_list[[1]]
```

### Joint normalization of Hi-C datasets

First, we need to create a `Hicexp` object using the Hi-C data:

```{r}
# Create a Hicexp object for use by multiHiCcompare (~10 min)
# Four objects are assigned into two groups
rao2017 <- make_hicexp(data_list = sample_list, groups = c(1, 1, 2, 2))

rao2017 # class(rao2017)
```

The `Hicexp` object stores the Hi-C experiment data and is the main input into the other functions included in `multiHiCcompare`. The user can view the IF information by using the `hic_table` accessor function:

```{r}
hic_table(rao2017)
```

When comparing multiple Hi-C datasets, a joint normalization procedure increases power and reduces the number of false positives [@Stansfield2018]. The `multiHiCcompare` R package includes two methods for the joint normalization of Hi-C data, cyclic loess and fast loess (fastlo) [@Ballman2004]. We will normalize the data using fastlo.

```{r}
# MD plots before normalization
MD_hicexp(rao2017, plot.chr = 1, plot.loess = TRUE)
```

```{r, results="hide", echo = FALSE, eval = FALSE}
# MD plots before normalization
png(filename = "inst/vignettes/fig1.png", width = 1500, height = 1500, units = 'px', res = 150)
MD_hicexp(rao2017, plot.chr = 1, plot.loess = TRUE)
dev.off()
```


```{r}
# Normalize (~2 min)
rao2017 <- fastlo(rao2017)
```

```{r, eval = FALSE}
# Plot normalization results
MD_hicexp(rao2017, plot.chr = 1, plot.loess = TRUE)
```

```{r, results="hide", echo = FALSE, eval = FALSE}
# Plot normalization results
tiff(filename = "figures/fig3.tiff", width = 1500, height = 1500, units = 'px', res = 300)
MD_hicexp(rao2017, plot.chr = 1, plot.loess = TRUE, prow = 1, pcol = 1)
dev.off()
```


```{r}
# Print normalized IFs
knitr::kable(head(hic_table(rao2017)))
```

The IFs in the `hic_table` slot have been updated with their normalized values. The MD plots (Fig. 2) show that the normalization has been performed correctly, and the cloud of points is centered and symmetric around 0 indicating that any biases between datasets have been removed. The MD plot displays unit genomic distance on the x-axis and the log2 difference between the two datasets on the y-axis. Any shift of the points away from y = 0 represents scaling differences between the datasets. The loess fit to the data on the MD plot will also model any trend biases between the datasets. Correctly normalized data should be centered around y = 0 and symmetric (without any clear trends) on the MD plot.

Note that if multiple cores are available, the runtime of `multiHiCcompare` can be sped up by using the `parallel` option. `multiHiCcompare` was built with the Bioconductor `BiocParallel` package. The number of cores to be used in parallel processing can be set as follows:

```{r}
library(BiocParallel) # BiocManager::install("BiocParallel")
# Check how many cores are available
numCores <- parallel::detectCores()

# Set the number of cores at least one less than the total number
if(Sys.info()['sysname'] == "Windows") {
  # Windows settings
  register(SnowParam(workers = numCores - 1), default = TRUE)
} else {
  # Unix settings
  register(MulticoreParam(workers = numCores - 1), default = TRUE)
}
```

Now that multiple cores are registered, the user can utilize parallel processing in any of the normalization and difference detection steps by setting `parallel = TRUE` in the function options.



### Difference detection

Now that we have jointly normalized our data, we are ready to compare the conditions to find differentially interacting chromatin regions. For this example, we only have two conditions and no other covariates. Thus, we can use the exact test for our comparison:

```{r}
# Perform exact test (~10 min)
# May use "parallel = TRUE" option to speed up computations
rao2017 <- hic_exactTest(rao2017, parallel = TRUE) 
```



```{r}
# Plot a composite MD plot with the results of a comparison
MD_composite(rao2017, plot.chr = 1)
```

```{r, results="hide", echo = FALSE, eval = FALSE}
# Plot a composite MD plot with the results of a comparison
tiff(filename = "figures/fig4.tiff", width = 1500, height = 1500, units = 'px', res = 300)
MD_composite(rao2017, plot.chr = 1)
dev.off()
```


```{r}
# Print results as a data frame
knitr::kable(head(results(rao2017)))

# Save the Hicexp object
save(rao2017, file = 'rao2017.RDA')

# To start the downstream analysis 
# without re-running multiHiCcompare load the saved file
# load('rao2017.RDA') 
```

Here we can see the results. The composite MD plot highlights where the significantly different interactions are occurring in relation to distance and the fold change of the difference between groups (Fig. 3). The results table shares the same first four columns with the `hic_table`, but the following columns indicate the results of the exact test. `logFC` is the log fold change difference between the experimental groups, `logCPM` is the log counts per million between the samples, `p.value` is the un-adjusted p-value for the exact test, and `p.adj` is the false discovery rate (FDR) corrected p-value from the exact test. 

