<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title> â€¢ HiCcompareWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="">
<meta property="og:description" content="HiCcompareWorkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">HiCcompareWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/hic_tutorial.html">UNKNOWN TITLE</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mdozmorov/HiCcompareWorkshop/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip></h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mdozmorov/HiCcompareWorkshop/blob/master/vignettes/hic_tutorial.Rmd"><code>vignettes/hic_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>hic_tutorial.Rmd</code></div>

    </div>

    
    
<div id="detection-of-differentially-interacting-chromatin-regions-from-multiple-hi-c-datasets" class="section level1">
<h1 class="hasAnchor">
<a href="#detection-of-differentially-interacting-chromatin-regions-from-multiple-hi-c-datasets" class="anchor"></a>Detection of Differentially Interacting Chromatin Regions From Multiple Hi-C Datasets</h1>
<div id="workshop-information" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-information" class="anchor"></a>Workshop information</h2>
<div id="instructors-names-and-contact-information" class="section level3">
<h3 class="hasAnchor">
<a href="#instructors-names-and-contact-information" class="anchor"></a>Instructor(s) name(s) and contact information</h3>
<p>Mikhail Dozmorov</p>
<p>Department of Biostatistics<br>
Virginia Commonwealth University<br>
830 E. Main St<br>
Richmond, VA, 23298<br><a href="mailto:mikhail.dozmorov@vcuhealth.org" class="email">mikhail.dozmorov@vcuhealth.org</a></p>
</div>
<div id="workshop-description" class="section level3">
<h3 class="hasAnchor">
<a href="#workshop-description" class="anchor"></a>Workshop Description</h3>
<p>This tutorial will introduce methods for the comparative (aka differential) analysis of the three-dimensional (3D) structure of the genome using data generated by high throughput chromatin conformation capture (Hi-C) technologies. Hi-C data allows for insights into the genome-wide genomic interactions which play an important role in the regulation of the genome. Just as differential expression analyses using RNA-seq data have become a routine part of genomic experiments, we expect the differential analysis of genomic interactions to become a common task for a bioinformatician. This workflow will help a novice learn to perform differential analysis of two or more Hi-C datasets and interpret the results of the differential genomic interactions.</p>
</div>
<div id="pre-requisites" class="section level3">
<h3 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h3>
<ul>
<li>Basic knowledge of R syntax and command-line tools</li>
<li>Familiarity with Hi-C chromatin conformation capture technology and FASTQ format</li>
<li>Understanding of Hi-C data properties (distance-dependent decay of interaction frequencies, biases)</li>
<li>Familiarity with specialized (<code>.hic</code> <a href="http://aidenlab.org/data.html">http://aidenlab.org/data.html</a>, <code>.cool</code> <a href="ftp://cooler.csail.mit.edu/coolers">ftp://cooler.csail.mit.edu/coolers</a>) and text-based (sparse upper-triangular, full square matrix) Hi-C data formats</li>
</ul>
</div>
<div id="workshop-participation" class="section level3">
<h3 class="hasAnchor">
<a href="#workshop-participation" class="anchor"></a>Workshop Participation</h3>
<p>Execution of example code and hands-on practice</p>
</div>
<div id="r-bioconductor-packages-used" class="section level3">
<h3 class="hasAnchor">
<a href="#r-bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h3>
<ul>
<li><a href="https://www.bioconductor.org/packages/HiCcompare">HiCcompare</a></li>
<li><a href="https://bioconductor.org/packages/multiHiCcompare">multiHiCcompare</a></li>
</ul>
</div>
<div id="time-outline" class="section level3">
<h3 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h3>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Overview</td>
<td>5m</td>
</tr>
<tr class="even">
<td>Data representation and manipulation</td>
<td>20m</td>
</tr>
<tr class="odd">
<td>Differential analysis of Hi-C data</td>
<td>20m</td>
</tr>
<tr class="even">
<td>Interpretation of Hi-C differences</td>
<td>15m</td>
</tr>
</tbody>
</table>
</div>
<div id="learning-goals" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-goals" class="anchor"></a>Learning Goals</h3>
<ul>
<li>Get familiar with Hi-C data import into R</li>
<li>Understand the visualization of biases between pairs of Hi-C datasets using Mean-Distance (MD) plot</li>
<li>Learn the Loess-based normalization strategy that minimizes between-dataset differences</li>
<li>Perform differential analysis of 2X2 groups of Hi-C data</li>
<li>Visualize and understand interaction frequency differences</li>
<li>Learn how the significance of differential interactions changes with the increasing distance between interacting regions</li>
<li>Get familiar with approaches to interpret chromatin interaction differences in context of genes</li>
</ul>
</div>
<div id="working-with-processed-hi-c-data" class="section level3">
<h3 class="hasAnchor">
<a href="#working-with-processed-hi-c-data" class="anchor"></a>Working with processed Hi-C data</h3>
<p>The <code>HiCcompare</code> R package was designed for working with processed Hi-C data. It contains several functions which may be useful for an analysis. Hi-C data extracted from <code>.hic</code> files using <code>straw</code> can be simply read into R in the standard fashion for loading any text file containing data. <code>HiCcompare</code> can then be used to convert Hi-C data from sparse upper triangular matrix format into a full contact matrix using the <code>sparse2full</code> function. This process can be reversed using the <code>full2sparse</code> function. Data aligned by <code>HiC-Pro</code> can be converted into a more usable <code>BEDPE</code> format using the <code>hicpro2bedpe</code> function. The <code>hicpro2bedpe</code> function takes the <code>.matrix</code> and <code>.bed</code> files produced by <code>HiC-Pro</code> as input and produces a sparse upper triangular matrix containing start and end coordinates for each interacting region.</p>
<p>The original <code>HiCCompare</code> R package can be used when only two Hi-C datasets are available to be compared <span class="citation">[@Stansfield2018]</span>. <code>HiCcompare</code> provides a method for the joint normalization and difference detection of two Hi-C datasets, but cannot be generalized to higher numbers of datasets. <code>multiHiCcompare</code> will need to be used if more than two Hi-C datasets are to be compared <span class="citation">[@Stansfield:multihiccompare]</span>.</p>
</div>
<div id="obtaining-and-preparing-the-data" class="section level3">
<h3 class="hasAnchor">
<a href="#obtaining-and-preparing-the-data" class="anchor"></a>Obtaining and preparing the data</h3>
<p>We begin our tutorial illustrating the processing of Hi-C datasets to prepare them for comparative analysis. First, we will need to download an example set of Hi-C data. We will use data from Rao 2017 <span class="citation">[@Rao:2017aa]</span>. For simplicity, we will only use two replicates for each experimental condition. The experimental conditions are normal HCT-116 cells and HCT-116 cells treated with auxin for six hours. To download the <code>.hic</code> files from GEO run the following commands in the terminal. Note: downloading the data for this protocol will require about 30GB of hard drive space. This step is for illustration purposes only, we will work with the much smaller data already processed using these commands.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1"></a><span class="fu">wget</span> https://github.com/theaidenlab/straw/tree/master/bin/Mac/straw</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="fu">chmod</span> +x straw</span>
<span id="cb1-3"><a href="#cb1-3"></a></span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="fu">wget</span> ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM2795nnn/GSM2795535/suppl/GSM2795535_Rao-2017-HIC001_30.hic</span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="fu">wget</span> ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM2795nnn/GSM2795536/suppl/GSM2795536_Rao-2017-HIC002_30.hic</span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="fu">wget</span> ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM2809nnn/GSM2809539/suppl/GSM2809539_Rao-2017-HIC008_30.hic</span>
<span id="cb1-7"><a href="#cb1-7"></a><span class="fu">wget</span> ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM2809nnn/GSM2809540/suppl/GSM2809540_Rao-2017-HIC009_30.hic</span>
<span id="cb1-8"><a href="#cb1-8"></a></span>
<span id="cb1-9"><a href="#cb1-9"></a><span class="co"># Make directories for the contact map files</span></span>
<span id="cb1-10"><a href="#cb1-10"></a><span class="fu">mkdir</span> HIC001</span>
<span id="cb1-11"><a href="#cb1-11"></a><span class="fu">mkdir</span> HIC002</span>
<span id="cb1-12"><a href="#cb1-12"></a><span class="fu">mkdir</span> HIC008</span>
<span id="cb1-13"><a href="#cb1-13"></a><span class="fu">mkdir</span> HIC009</span>
<span id="cb1-14"><a href="#cb1-14"></a></span>
<span id="cb1-15"><a href="#cb1-15"></a><span class="co"># Extract contact maps using straw by running the following commands in the terminal</span></span>
<span id="cb1-16"><a href="#cb1-16"></a><span class="co"># Or, put the commands into a script file, e.g., `straw.sh`, and run it</span></span>
<span id="cb1-17"><a href="#cb1-17"></a><span class="kw">for</span> <span class="ex">i</span> in <span class="dt">{1..22}</span></span>
<span id="cb1-18"><a href="#cb1-18"></a>    <span class="kw">do</span></span>
<span id="cb1-19"><a href="#cb1-19"></a>        <span class="ex">./straw</span> NONE GSM2795535_Rao-2017-HIC001_30.hic <span class="va">$i</span> <span class="va">$i</span> BP 500000 <span class="op">&gt;</span> HIC001/HIC001.NONE.chr<span class="va">$i</span>.500000.txt</span>
<span id="cb1-20"><a href="#cb1-20"></a>    <span class="kw">done</span></span>
<span id="cb1-21"><a href="#cb1-21"></a>    <span class="ex">./straw</span> NONE GSM2795535_Rao-2017-HIC001_30.hic X X BP 500000 <span class="op">&gt;</span> HIC001/HIC001.NONE.chrX.500000.txt</span>
<span id="cb1-22"><a href="#cb1-22"></a>    </span>
<span id="cb1-23"><a href="#cb1-23"></a><span class="kw">for</span> <span class="ex">i</span> in <span class="dt">{1..22}</span></span>
<span id="cb1-24"><a href="#cb1-24"></a>    <span class="kw">do</span></span>
<span id="cb1-25"><a href="#cb1-25"></a>        <span class="ex">./straw</span> NONE GSM2795536_Rao-2017-HIC002_30.hic <span class="va">$i</span> <span class="va">$i</span> BP 500000 <span class="op">&gt;</span> HIC002/HIC002.NONE.chr<span class="va">$i</span>.500000.txt</span>
<span id="cb1-26"><a href="#cb1-26"></a>    <span class="kw">done</span></span>
<span id="cb1-27"><a href="#cb1-27"></a>    <span class="ex">./straw</span> NONE GSM2795536_Rao-2017-HIC002_30.hic X X BP 500000 <span class="op">&gt;</span> HIC002/HIC002.NONE.chrX.500000.txt</span>
<span id="cb1-28"><a href="#cb1-28"></a>    </span>
<span id="cb1-29"><a href="#cb1-29"></a><span class="kw">for</span> <span class="ex">i</span> in <span class="dt">{1..22}</span></span>
<span id="cb1-30"><a href="#cb1-30"></a>    <span class="kw">do</span></span>
<span id="cb1-31"><a href="#cb1-31"></a>        <span class="ex">./straw</span> NONE GSM2809539_Rao-2017-HIC008_30.hic <span class="va">$i</span> <span class="va">$i</span> BP 500000 <span class="op">&gt;</span> HIC008/HIC008.NONE.chr<span class="va">$i</span>.500000.txt</span>
<span id="cb1-32"><a href="#cb1-32"></a>    <span class="kw">done</span></span>
<span id="cb1-33"><a href="#cb1-33"></a>    <span class="ex">./straw</span> NONE GSM2809539_Rao-2017-HIC008_30.hic X X BP 500000 <span class="op">&gt;</span> HIC008/HIC008.NONE.chrX.500000.txt</span>
<span id="cb1-34"><a href="#cb1-34"></a>        </span>
<span id="cb1-35"><a href="#cb1-35"></a><span class="kw">for</span> <span class="ex">i</span> in <span class="dt">{1..22}</span></span>
<span id="cb1-36"><a href="#cb1-36"></a>    <span class="kw">do</span></span>
<span id="cb1-37"><a href="#cb1-37"></a>        <span class="ex">./straw</span> NONE GSM2809540_Rao-2017-HIC009_30.hic <span class="va">$i</span> <span class="va">$i</span> BP 500000 <span class="op">&gt;</span> HIC009/HIC009.NONE.chr<span class="va">$i</span>.500000.txt</span>
<span id="cb1-38"><a href="#cb1-38"></a>    <span class="kw">done</span></span>
<span id="cb1-39"><a href="#cb1-39"></a>    <span class="ex">./straw</span> NONE GSM2809540_Rao-2017-HIC009_30.hic X X BP 500000 <span class="op">&gt;</span> HIC009/HIC009.NONE.chrX.500000.txt</span></code></pre></div>
<p>These steps will create four folders containing the sparse upper triangular matrices for chromosomes 1-22, and X for each sample. HIC001 and HIC002 are the two replicates for the normal HCT-116 cells and HIC008 and HIC009 are the two replicates for the auxin-treated HCT-116 cells.</p>
<p>We now need to read the data into R. Open R and make sure the working directory is set to the directory where the Hi-C data is stored and execute the following commands:</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="co"># Install, if necessary, and load necessary libraries and set up R session</span>
<span class="kw">if</span> (!<span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"BiocManager"</span>, <span class="kw">quietly</span> <span class="kw">=</span> <span class="fl">TRUE</span>)) <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">readr</span>) <span class="co"># install.packages("readr")</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">data.table</span>) <span class="co"># install.packages("data.table")</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dplyr</span>) <span class="co"># install.packages("dplyr")</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">edgeR</span>) <span class="co"># BiocManager::install("edgeR")</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BiocParallel</span>) <span class="co"># BiocManager::install("BiocParallel") </span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">HiCcompare</span>) <span class="co"># BiocManager::install("HiCcompare"), or, for the latest version, devtools::install_github('dozmorovlab/HiCcompare', build_vignettes = TRUE, force = TRUE)</span>
<span class="co"># install.packages("devtools")</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">multiHiCcompare</span>) <span class="co"># BiocManager::install("multiHiCcompare", version = "devel") </span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="kw">scipen</span> <span class="kw">=</span> <span class="fl">10</span>) <span class="co"># Output fixed numbers, not scientific notation</span>

<span class="co"># Set up parameters for reading in data</span>
<span class="no">chr</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">'chr'</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>:<span class="fl">22</span>, <span class="st">'X'</span>)) <span class="co"># Chromosome names</span>
<span class="no">chr</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">'chr'</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>)) <span class="co"># Chromosome names</span>
<span class="no">dirIn</span> <span class="kw">&lt;-</span> <span class="st">"data/"</span> <span class="co"># Input folder</span>
<span class="no">dirIn</span> <span class="kw">&lt;-</span> <span class="st">"https://github.com/kmt555/ComparativeHiC-analysis-tutorial/raw/master/data/"</span> <span class="co"># Remote data</span>
<span class="no">samples</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">'HIC00'</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">8</span>, <span class="fl">9</span>)) <span class="co"># Sample names</span>
<span class="no">res</span> <span class="kw">&lt;-</span> <span class="fl">500000</span> <span class="co"># Data resolution</span>

<span class="co"># Read data</span>
<span class="no">sample_list</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>()
<span class="no">chr_list</span>    <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>()
<span class="kw">for</span>( <span class="no">j</span> <span class="kw">in</span> <span class="fl">1</span>:<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">samples</span>)) {
  <span class="kw">for</span> (<span class="no">i</span> <span class="kw">in</span> <span class="fl">1</span>:<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">chr</span>)) {
    <span class="no">chr_list</span><span class="kw">[[</span><span class="no">i</span>]] <span class="kw">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_tsv</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="no">dirIn</span>, <span class="no">samples</span>[<span class="no">j</span>], <span class="st">"/"</span>, <span class="no">samples</span>[<span class="no">j</span>],
                                     <span class="st">".NONE."</span>, <span class="no">chr</span>[<span class="no">i</span>], <span class="st">"."</span>, <span class="no">res</span>, <span class="st">".txt.gz"</span>),
                              <span class="kw">col_names</span> <span class="kw">=</span> <span class="fl">FALSE</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/as.data.table.html">as.data.table</a></span>()
    <span class="co"># Add column indicating the chromosome</span>
    <span class="no">chr_list</span><span class="kw">[[</span><span class="no">i</span>]] <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cbind.html">cbind</a></span>(<span class="no">i</span>, <span class="no">chr_list</span><span class="kw">[[</span><span class="no">i</span>]])
    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">chr_list</span><span class="kw">[[</span><span class="no">i</span>]]) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'chr'</span>, <span class="st">'region1'</span>, <span class="st">'region2'</span>, <span class="st">'IF'</span>)
  }
  <span class="no">sample_list</span><span class="kw">[[</span><span class="no">j</span>]] <span class="kw">&lt;-</span> <span class="no">chr_list</span>
  <span class="no">chr_list</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>()
}

<span class="co"># Collapse separate chromosome lists into one table per sample</span>
<span class="no">sample_list</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">sample_list</span>, <span class="no">rbindlist</span>)</pre></body></html></div>
<p>We now have a list with each entry containing the sparse upper triangular matrix for one of the Hi-C datasets:</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">sample_list</span><span class="kw">[[</span><span class="fl">1</span>]]</pre></body></html></div>
<pre><code>        chr   region1   region2    IF
     1:   1         0         0    17
     2:   1    500000    500000 11583
     3:   1    500000   1000000  5454
     4:   1   1000000   1000000 30858
     5:   1    500000   1500000   868
    ---                              
102703:   1 247000000 249000000  1166
102704:   1 247500000 249000000  1138
102705:   1 248000000 249000000   823
102706:   1 248500000 249000000  2043
102707:   1 249000000 249000000  8336</code></pre>
</div>
<div id="joint-normalization-of-hi-c-datasets" class="section level3">
<h3 class="hasAnchor">
<a href="#joint-normalization-of-hi-c-datasets" class="anchor"></a>Joint normalization of Hi-C datasets</h3>
<p>First, we need to create a <code>Hicexp</code> object using the Hi-C data:</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="co"># Create a Hicexp object for use by multiHiCcompare (~10 min)</span>
<span class="co"># Four objects are assigned into two groups</span>
<span class="no">rao2017</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/multiHiCcompare/man/make_hicexp.html">make_hicexp</a></span>(<span class="kw">data_list</span> <span class="kw">=</span> <span class="no">sample_list</span>, <span class="kw">groups</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span>))

<span class="no">rao2017</span> <span class="co"># class(rao2017)</span></pre></body></html></div>
<pre><code>Hi-C Experiment Object 
2 experimental groups 
Group 1 has 2 samples 
Group 2 has 2 samples </code></pre>
<p>The <code>Hicexp</code> object stores the Hi-C experiment data and is the main input into the other functions included in <code>multiHiCcompare</code>. The user can view the IF information by using the <code>hic_table</code> accessor function:</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/multiHiCcompare/man/hic_table.html">hic_table</a></span>(<span class="no">rao2017</span>)</pre></body></html></div>
<pre><code>       chr   region1   region2 D   IF1   IF2  IF3  IF4
    1:   1         0         0 0    17    12    5   13
    2:   1    500000    500000 0 11583 16945 4694 6252
    3:   1    500000   1000000 1  5454  7462 1942 2475
    4:   1    500000   1500000 2   868  1103  720  856
    5:   1    500000   2000000 3   480   633  521  647
   ---                                                
95549:   1 248000000 248500000 1  4009  3839 2290 2333
95550:   1 248000000 249000000 2   823   856  507  533
95551:   1 248500000 248500000 0  8283  8848 5391 5892
95552:   1 248500000 249000000 1  2043  2178  848  950
95553:   1 249000000 249000000 0  8336 10546 4272 5359</code></pre>
<p>When comparing multiple Hi-C datasets, a joint normalization procedure increases power and reduces the number of false positives <span class="citation">[@Stansfield2018]</span>. The <code>multiHiCcompare</code> R package includes two methods for the joint normalization of Hi-C data, cyclic loess and fast loess (fastlo) <span class="citation">[@Ballman2004]</span>. We will normalize the data using fastlo.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="co"># MD plots before normalization</span>
<span class="fu"><a href="https://rdrr.io/pkg/multiHiCcompare/man/MD_hicexp.html">MD_hicexp</a></span>(<span class="no">rao2017</span>, <span class="kw">plot.chr</span> <span class="kw">=</span> <span class="fl">1</span>, <span class="kw">plot.loess</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<p><img src="img/unnamed-chunk-6-1.png" width="1000"></p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="co"># Normalize (~2 min)</span>
<span class="no">rao2017</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/multiHiCcompare/man/fastlo.html">fastlo</a></span>(<span class="no">rao2017</span>)</pre></body></html></div>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="co"># Plot normalization results</span>
<span class="fu"><a href="https://rdrr.io/pkg/multiHiCcompare/man/MD_hicexp.html">MD_hicexp</a></span>(<span class="no">rao2017</span>, <span class="kw">plot.chr</span> <span class="kw">=</span> <span class="fl">1</span>, <span class="kw">plot.loess</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="co"># Print normalized IFs</span>
<span class="kw pkg">knitr</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span>(<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/multiHiCcompare/man/hic_table.html">hic_table</a></span>(<span class="no">rao2017</span>)))</pre></body></html></div>
<table class="table">
<thead><tr class="header">
<th align="right">chr</th>
<th align="right">region1</th>
<th align="right">region2</th>
<th align="right">D</th>
<th align="right">IF1</th>
<th align="right">IF2</th>
<th align="right">IF3</th>
<th align="right">IF4</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">14.03845</td>
<td align="right">9.054028</td>
<td align="right">6.583487</td>
<td align="right">16.14285</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="right">500000</td>
<td align="right">500000</td>
<td align="right">0</td>
<td align="right">7976.91924</td>
<td align="right">10874.195738</td>
<td align="right">7464.948223</td>
<td align="right">8895.88720</td>
</tr>
<tr class="odd">
<td align="right">1</td>
<td align="right">500000</td>
<td align="right">1000000</td>
<td align="right">1</td>
<td align="right">3735.70219</td>
<td align="right">4822.706786</td>
<td align="right">3074.815893</td>
<td align="right">3531.66059</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="right">500000</td>
<td align="right">1500000</td>
<td align="right">2</td>
<td align="right">784.64047</td>
<td align="right">939.231570</td>
<td align="right">866.952110</td>
<td align="right">923.59278</td>
</tr>
<tr class="odd">
<td align="right">1</td>
<td align="right">500000</td>
<td align="right">2000000</td>
<td align="right">3</td>
<td align="right">583.88128</td>
<td align="right">727.500686</td>
<td align="right">467.083320</td>
<td align="right">516.20057</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="right">500000</td>
<td align="right">2500000</td>
<td align="right">4</td>
<td align="right">283.38905</td>
<td align="right">353.270921</td>
<td align="right">264.837459</td>
<td align="right">258.98222</td>
</tr>
</tbody>
</table>
<p>The IFs in the <code>hic_table</code> slot have been updated with their normalized values. The MD plots (Fig. 2) show that the normalization has been performed correctly, and the cloud of points is centered and symmetric around 0 indicating that any biases between datasets have been removed. The MD plot displays unit genomic distance on the x-axis and the log2 difference between the two datasets on the y-axis. Any shift of the points away from y = 0 represents scaling differences between the datasets. The loess fit to the data on the MD plot will also model any trend biases between the datasets. Correctly normalized data should be centered around y = 0 and symmetric (without any clear trends) on the MD plot.</p>
<p>Note that if multiple cores are available, the runtime of <code>multiHiCcompare</code> can be sped up by using the <code>parallel</code> option. <code>multiHiCcompare</code> was built with the Bioconductor <code>BiocParallel</code> package. The number of cores to be used in parallel processing can be set as follows:</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BiocParallel</span>) <span class="co"># BiocManager::install("BiocParallel")</span>
<span class="co"># Check how many cores are available</span>
<span class="no">numCores</span> <span class="kw">&lt;-</span> <span class="kw pkg">parallel</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html">detectCores</a></span>()

<span class="co"># Set the number of cores at least one less than the total number</span>
<span class="kw">if</span>(<span class="fu"><a href="https://rdrr.io/r/base/Sys.info.html">Sys.info</a></span>()[<span class="st">'sysname'</span>] <span class="kw">==</span> <span class="st">"Windows"</span>) {
  <span class="co"># Windows settings</span>
  <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html">register</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html">SnowParam</a></span>(<span class="kw">workers</span> <span class="kw">=</span> <span class="no">numCores</span> - <span class="fl">1</span>), <span class="kw">default</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
} <span class="kw">else</span> {
  <span class="co"># Unix settings</span>
  <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html">register</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html">MulticoreParam</a></span>(<span class="kw">workers</span> <span class="kw">=</span> <span class="no">numCores</span> - <span class="fl">1</span>), <span class="kw">default</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
}</pre></body></html></div>
<p>Now that multiple cores are registered, the user can utilize parallel processing in any of the normalization and difference detection steps by setting <code>parallel = TRUE</code> in the function options.</p>
</div>
<div id="difference-detection" class="section level3">
<h3 class="hasAnchor">
<a href="#difference-detection" class="anchor"></a>Difference detection</h3>
<p>Now that we have jointly normalized our data, we are ready to compare the conditions to find differentially interacting chromatin regions. For this example, we only have two conditions and no other covariates. Thus, we can use the exact test for our comparison:</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="co"># Perform exact test (~10 min)</span>
<span class="co"># May use "parallel = TRUE" option to speed up computations</span>
<span class="no">rao2017</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/multiHiCcompare/man/hic_exactTest.html">hic_exactTest</a></span>(<span class="no">rao2017</span>, <span class="kw">parallel</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="co"># Plot a composite MD plot with the results of a comparison</span>
<span class="fu"><a href="https://rdrr.io/pkg/multiHiCcompare/man/MD_composite.html">MD_composite</a></span>(<span class="no">rao2017</span>, <span class="kw">plot.chr</span> <span class="kw">=</span> <span class="fl">1</span>)</pre></body></html></div>
<p><img src="img/unnamed-chunk-14-1.png" width="1000"></p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="co"># Print results as a data frame</span>
<span class="kw pkg">knitr</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span>(<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/multiHiCcompare/man/results.html">results</a></span>(<span class="no">rao2017</span>)))</pre></body></html></div>
<table class="table">
<thead><tr class="header">
<th align="right">chr</th>
<th align="right">region1</th>
<th align="right">region2</th>
<th align="right">D</th>
<th align="right">logFC</th>
<th align="right">logCPM</th>
<th align="right">p.value</th>
<th align="right">p.adj</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">-0.0174526</td>
<td align="right">-0.0045321</td>
<td align="right">1.0000000</td>
<td align="right">1.0000000</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="right">500000</td>
<td align="right">500000</td>
<td align="right">0</td>
<td align="right">-0.2778767</td>
<td align="right">9.3487766</td>
<td align="right">0.2724635</td>
<td align="right">0.3478203</td>
</tr>
<tr class="odd">
<td align="right">1</td>
<td align="right">500000</td>
<td align="right">1000000</td>
<td align="right">1</td>
<td align="right">-0.5220235</td>
<td align="right">8.1336596</td>
<td align="right">0.0167912</td>
<td align="right">0.0329749</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="right">500000</td>
<td align="right">1500000</td>
<td align="right">2</td>
<td align="right">0.0952323</td>
<td align="right">6.0276831</td>
<td align="right">0.5575227</td>
<td align="right">0.6320454</td>
</tr>
<tr class="odd">
<td align="right">1</td>
<td align="right">500000</td>
<td align="right">2000000</td>
<td align="right">3</td>
<td align="right">-0.5944782</td>
<td align="right">8.6530833</td>
<td align="right">0.0008158</td>
<td align="right">0.0033206</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="right">500000</td>
<td align="right">2500000</td>
<td align="right">4</td>
<td align="right">-0.4011353</td>
<td align="right">7.6744958</td>
<td align="right">0.0319779</td>
<td align="right">0.0680869</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="co"># Save the Hicexp object</span>
<span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span>(<span class="no">rao2017</span>, <span class="kw">file</span> <span class="kw">=</span> <span class="st">'rao2017.RDA'</span>)

<span class="co"># To start the downstream analysis </span>
<span class="co"># without re-running multiHiCcompare load the saved file</span>
<span class="co"># load('rao2017.RDA') </span></pre></body></html></div>
<p>Here we can see the results. The composite MD plot highlights where the significantly different interactions are occurring in relation to distance and the fold change of the difference between groups (Fig. 3). The results table shares the same first four columns with the <code>hic_table</code>, but the following columns indicate the results of the exact test. <code>logFC</code> is the log fold change difference between the experimental groups, <code>logCPM</code> is the log counts per million between the samples, <code>p.value</code> is the un-adjusted p-value for the exact test, and <code>p.adj</code> is the false discovery rate (FDR) corrected p-value from the exact test.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Mikhail Dozmorov.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
